api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java

```diff
@@ -69,13 +71,26 @@ public StrictMetricsEvaluator(Schema schema, Expression unbound, boolean caseSen
    *     otherwise.
    */
   public boolean eval(ContentFile<?> file) {
-    // TODO: detect the case where a column is missing from the file using file's max field id.
+    if (file.valueCounts() != null) {
+      int maxFieldId = file.valueCounts().keySet().stream().mapToInt(i -> i).max().orElse(0);
```

Comment:
I'm not sure this way of maxFieldId from the `ContentFile` is correct. Isn't it possible that `valueCounts` stats are missing stats for higher field IDs that are in the file? It's not a guarantee that all the field IDs are in this map.

Reply:
I agree with Amogh here, and we cannot use the `maxFieldId`. For example, the [default write configuration](https://iceberg.apache.org/docs/nightly/configuration/#write-properties) has `write.metadata.metrics.max-inferred-column-defaults` where we limit the collection of statistics to 100 columns by default. This would already break this approach as if we would filter on a column that comes after the first 100 columns, then it would already short-circuit and return an empty result.

Reply:
@amogh-jahagirdar @Fokko Thanks for your suggestions. Will re-work based on your inputs.

Reply:
Thanks for taking the suggestion into consideration. Since you already have the schema, you could also build a set of the field IDs to check if the column is missing.

Keep in mind that not all the Parquet files have the field IDs set. If you convert an existing Hive table into an Iceberg table, the Iceberg will leverage [name-mapping](https://iceberg.apache.org/spec/#column-projection) to map the names into an Iceberg field-ID.

Reply:
Schema might have 10 columns but the file could have 5 columns (say, field id from 1 to 5). So, we need to set 5 as max field id on that specific file. Then only, we can able to use this max field id to skip the files who's max field id is lesser than id of the column come through `Expression` to *MetricsEvaluator. Is my understanding correct?

After writing the file using `DataWriter`, list of columns could be fetched from getFileMetaData().getSchema() and max value from the same can be set on `DataFile` or `ContentFile`. Which, in turn, could be used to skip the file as described earlier.

Thoughts?

Reply:
@Fokko Pinging you to do interim review to ensure direction is right.